lasy/profiles/speckle_profile.py

```diff
@@ -0,0 +1,519 @@
+        sampled from a Gaussian stochastic process.
+
+    relative_laser_bandwidth : float
+        Bandwidth of laser pulse, relative to central frequency.
+
+    ssd_phase_modulation_amplitude : 2-tuple of floats
+        Amplitude of phase modulation in each transverse direction.
+        Only used if `temporal_smoothing_type` is `FM SSD`.
+
+    ssd_number_color_cycles : list of 2 floats
+        Number of color cycles of SSD spectrum to include in modulation
+        Only used if `temporal_smoothing_type` is `FM SSD`.
+
+    ssd_transverse_bandwidth_distribution: list of 2 floats
+        Determines how much SSD is distributed in the `x` and `y` directions.
+        if `ssd_transverse_bandwidth_distribution=[a,b]`, then the SSD frequency modulation is `a/sqrt(a^2+b^2)` in `x` and `b/sqrt(a^2+b^2)` in `y`.
+        Only used if `temporal_smoothing_type` is `FM SSD`.
+
+    do_include_transverse_envelope : boolean, (optional, default False)
+        Whether to include the transverse sinc envelope or not.
+        I.e. whether it is assumed to be close enough to the laser axis to neglect the transverse field decay.
+    """
+
+    supported_smoothing = "RPP", "CPP", "FM SSD", "GP RPM SSD", "GP ISI"
+
+    def __init__(
+        self,
+        wavelength,
+        pol,
+        focal_length,
+        beam_aperture,
+        n_beamlets,
+        temporal_smoothing_type,
+        relative_laser_bandwidth,
+        ssd_phase_modulation_amplitude=None,
+        ssd_number_color_cycles=None,
+        ssd_transverse_bandwidth_distribution=None,
+        do_include_transverse_envelope=False,
+    ):
+        super().__init__(wavelength, pol)
+        self.wavelength = wavelength
+        self.focal_length = focal_length
+        self.beam_aperture = np.array(beam_aperture, dtype="float")
+        self.n_beamlets = np.array(n_beamlets, dtype="int")
+        self.temporal_smoothing_type = temporal_smoothing_type
+        self.laser_bandwidth = relative_laser_bandwidth
+
+        # time interval to update the speckle pattern, roughly update 50 times every bandwidth cycle
+        self.dt_update = 1 / self.laser_bandwidth / 50
+        self.do_include_transverse_envelope = do_include_transverse_envelope
+
+        self.x_lens_list = np.linspace(
+            -0.5 * (self.n_beamlets[0] - 1),
+            0.5 * (self.n_beamlets[0] - 1),
+            num=self.n_beamlets[0],
+        )
+        self.y_lens_list = np.linspace(
+            -0.5 * (self.n_beamlets[1] - 1),
+            0.5 * (self.n_beamlets[1] - 1),
+            num=self.n_beamlets[1],
+        )
+        self.Y_lens_matrix, self.X_lens_matrix = np.meshgrid(
+            self.y_lens_list, self.x_lens_list
+        )
+        self.Y_lens_index_matrix, self.X_lens_index_matrix = np.meshgrid(
+            np.arange(self.n_beamlets[1], dtype=float),
+            np.arange(self.n_beamlets[0], dtype=float),
+        )
+        self.set_phase_plate_phase_modulation()
```

Comment:
I am not sure that this needs to be a separate method (or that it needs to be called here)